Improve chinese module

merrickluo opened the pull request:
<!-- ⚠️ Please do not ignore this template! -->
I've made some improvements to the Chinese module and has been using it for years, here is what I add

- added a `+childframe` feature flag to make `pyim` show candidates in a childframe (using posframe).
- added a `+rime` feature flag to make pyim use [librime](https://github.com/rime/librime) as input method engine (new package [liberime](https://github.com/merrickluo/liberime)).
- use `pyim-cregexp-build` to make chinese characters can be searched with pinyin, similar to #6356 but without adding new package.
- disable `evil-snipe` when input method is activated, it intercepts key event that can mess up the input method.

-----
- [x] I searched the issue tracker and this hasn't been PRed before.
- [x] My changes are not on [the do-not-PR list](https://doomemacs.org/d/do-not-pr) for this project.
- [x] My commits conform to [the git conventions](https://doomemacs.org/d/git-conventions).
- [x] Any relevant issues or PRs have been linked to.

<!-- Remove checklist items above that don't apply to this PR -->

<!--

 ❤ Thank you for taking the time to contribute! Please be patient while we get
   around to reviewing your PR. 

   - Once a maintainer approves it, there's nothing left to do. It will
     eventually be merged.
   - If we convert your PR to a Draft, it means the verdict is undecided and we
     need more time to think about it.
   - If you decide to close your PR, please let us know why you did so.

-->

hlissner commented on modules/input/chinese/config.el:
```diff
@@ -6,7 +6,41 @@
+        ((featurep! :completion ivy)
+         (setq ivy-re-builders-alist
+               '((t . pyim-cregexp-ivy))))))
+
+(use-package! posframe
+  :when (featurep! +childframe))
+
+(use-package! liberime
+  :when (featurep! +rime)
+  :init
+  (setq liberime-auto-build t)
+  (setq liberime-user-data-dir (expand-file-name "rime" doom-cache-dir)))
+
+(use-package! pyim-liberime
+  :when (featurep! +rime)
+  :after liberime
+  :config
+  (setq pyim-default-scheme 'rime))
```
Please separate top-level `use-package` blocks with two blank lines.

hlissner commented on modules/input/chinese/config.el:
```diff
@@ -6,7 +6,41 @@
+         (setq ivy-re-builders-alist
+               '((t . pyim-cregexp-ivy))))))
+
+(use-package! posframe
+  :when (featurep! +childframe))
```
Lazy-load posframe. If pyim is expected to use it, use `:after pyim`.

hlissner commented on modules/input/chinese/packages.el:
```diff
@@ -5,3 +5,7 @@
+(when (featurep! +rime)
+  (package! liberime :pin "8291e22cd0990a99cb2f88ca67a9065a157f39af"))
+(when (featurep! +childframe)
+  (package! posframe))
```
Packages shouldn't be unpinned. Use `(package! posframe :pin "0d23bc5f7cfac00277d83ae7ba52c48685bcbc68")`

merrickluo replied:
pinded to latest commit. I see that multiple module has `(package! posframe :pin "")`, what would happen if they pin different versions?

hlissner replied:
Then the module that loads last will win. It's not ideal, but it's all we got at the moment.

merrickluo replied:
I see, thanks for clarifying.

hlissner commented on modules/input/chinese/config.el:
```diff
@@ -6,7 +6,41 @@
-        default-input-method "pyim"))
+        default-input-method "pyim")
+
+  (defun +input-method-activate-p ()
+    current-input-method)
```
Move this inside the `(after! evil-escape ...)` block, as it isn't used anywhere else.

hlissner commented on modules/input/chinese/config.el:
```diff
@@ -6,7 +6,41 @@
+
+  ;; allow vertico/selectrum search with pinyin
+  (cond ((featurep! :completion vertico)
+         (defadvice! +pinyin-orderless-regexp (orig-fn component)
+           :around 'orderless-regexp
+           (let ((result (funcall orig-fn component)))
+             (pyim-cregexp-build result))))
```
1. Use a `:filter-return` advice here instead.
2. Rename advice to `+pinyin-orderless-regexp-a` to conform to [naming conventions](https://docs.doomemacs.org/latest/#/developers/conventions/emacs-lisp/naming-conventions/doom-conventions).
3. Sharp quote function symbols (`#'orderless-regexp`).

```diff
  (cond ((featurep! :completion vertico)
-         (defadvice! +pinyin-orderless-regexp (orig-fn component)
-           :around 'orderless-regexp
-           (let ((result (funcall orig-fn component)))
-             (pyim-cregexp-build result))))
+         (defadvice! +pinyin-orderless-regexp-a (result)
+           :filter-return #'orderless-regexp
+           (pyim-cregexp-build result)))
         ((featurep! :completion ivy)
```

hlissner commented:
Can this PR salvage anything from #6482? I'd like to merge them into one, if possible. 

cc @merrickluo

merrickluo commented:
Thanks, @hlissner, I updated the PR addressing your comments, in your comments you mentioned #6382 which is the link to this PR, do you mean #6356 ?  
I see they use `pinyinlib` to achieve the same functionality as `pyim-cregexp-build`, so it's not needed, I'll try out the `evil-pinyin` package.

merrickluo commented:
I tried `evil-pinyin` and it's working great, so I copied the code from #6356 and made it only on with `evil +everywhere`.
Also, I see that the doc rewrite is now done, so I added some documentation about the new flags/packages.
cc @ErnestDong

ErnestDong commented:
thanks for your hard working

hlissner commented:
Sorry for the late response. I've merged this in separate parts to make few minor refactors, conformed the commit messages to our git conventions, and resolve some merge conflicts. Let me know if it gives you any trouble. Thanks for your help, in any case!

sdvcrx commented:
@hlissner  Hi! `doom sync` error if enable `chinese` module after pulling latest Doom Emacs config:

init.el:

```elisp
(doom! :input
       ;;bidi              ; (tfel ot) thgir etirw uoy gnipleh
       chinese
       ;;japanese
       ;;layout            ; auie,ctsrnm is the superior home row
;; ..........
```

First run `doom sync`: 

```
> Synchronizing "default" profile...
  > Regenerating envvars file
    ✓ Generated ~/.emacs.d/.local/env
  > Installing packages...
    > Updating recipe repos...
    > Cloning evil-pinyin... emacsmirror-mirror...
    - Checked out evil-pinyin: 3e9e501ded86f88e01a4edec5d526ab0fab879d7
    > Building evil-pinyin...
    ! Script was abruptly aborted, leaving Doom in an incomplete state!
    - Run 'doom sync' to repair it.
    ✓ Finished in 9.71810s
```

And run `doom sync` again:

```
> Synchronizing "default" profile...
  > Regenerating envvars file
    ✓ Generated ~/.emacs.d/.local/env
  x There was an unexpected runtime error
    Message: Symbol's function definition is void
    Details: (define-namespace)
    Backtrace:
      (define-namespace evil-pinyin- (defvar with-search-rule 'custom "Enabl...
      (load-with-code-conversion "/Users/sdvcrx/.emacs.d/.local/straight/bui...
      (load "/Users/sdvcrx/.emacs.d/.local/straight/build-29.0.60/evil-pinyi...
      (progn (load autoloads-file nil 'nomessage))
      (if (file-exists-p autoloads-file) (progn (load autoloads-file nil 'no...
      (let ((autoloads-file (straight--autoloads-file package))) (if (file-e...
      (straight--load-package-autoloads "evil-pinyin")
      (let ((name (symbol-name package))) (add-to-list 'load-path (directory...
      (let ((package (car tail))) (straight-register-package package) (let (...
      (while tail (let ((package (car tail))) (straight-register-package pac...
      (let ((tail (cl-delete-duplicates packages :test #'equal))) (while tai...
      (let (packages) (let ((tail doom-packages)) (while tail (let ((package...
    ! Wrote extended backtrace to ~/.emacs.d/.local/logs/cli.doom.230223180637.48268.error
  ! Script was abruptly aborted, leaving Doom in an incomplete state!
  - Run 'doom sync' to repair it.
```

My emacs version is 29 and with `native-comp`.

System Information: https://pastebin.com/smwcG0vc

merrickluo commented:
Thanks @hlissner ! I'm having the same issue as @sdvcrx, seems like `evil-pinyin` can't load `names` when compile. I didn't have this issue previously.

hlissner commented:
After looking into this, the issue runs deep. Possibly to do with Emacs 29's new loaddefs-gen library (which replaces autoload.el); a new library for generating autoloads files. In any case, since they aren't needed, I've disabled generation of evil-pinyin's autoloads file, which seems to work around this. Give it a try and let me know if that works.

merrickluo commented:
yeah, it syncs without a problem now,  `evil-pinyin`'s function works great as well. thanks @hlissner :bow:

sdvcrx commented:
Working perfectly, thanks @hlissner @merrickluo !

yangsheng6810 commented:
The following addition causes problem for users who enable both ivy and Chinese module. The command `pyim-cregexp-ivy` is defined in feature `pyim-cregexp-utils.el`, which [needs to be manually required](https://github.com/tumashu/pyim/blob/f70c4227752dfa53cd3922584624da2484346d53/README.org#%E8%AE%A9-ivy-%E6%94%AF%E6%8C%81%E6%8B%BC%E9%9F%B3%E6%90%9C%E7%B4%A2%E5%80%99%E9%80%89%E9%A1%B9%E5%8A%9F%E8%83%BD). If not required, ivy will try to use `pyim-cregexp-ivy` for all selection commands, which ends in error and makes Emacs almost unusable (try think of doing something without invoking any command that requires selection).

```emacs-lisp
(setq ivy-re-builders-alist
               '((t . pyim-cregexp-ivy)))
```
I currently workaround the issue with the following in my config file:
```emacs-lisp
(use-package pyim-cregexp-utils
  :defer t
  :commands (pyim-cregexp-ivy))
```
Unfortunately I am not familiar with how autoloading works in Doom Emacs, and it would be great if someone could make a fix in Doom itself.
